pkg/pool-weighted/contracts/WeightedPoolProtocolFees.sol

```diff
@@ -189,27 +189,23 @@ abstract contract WeightedPoolProtocolFees is BaseWeightedPool, ProtocolFeeCache
      * @dev Returns the amount of BPT to be minted to pay protocol fees on yield accrued by the Pool.
      * Note that this isn't a view function. This function automatically updates `_athRateProduct`  to ensure that
      * proper accounting is performed to prevent charging duplicate protocol fees.
+     * @return yieldProtocolFees - The amount of BPT to be minted as protocol fees on yield.
+     * @return athRateProduct - The new all-time-high rate product if it has increased, otherwise zero.
      */
     function _getYieldProtocolFee(uint256[] memory normalizedWeights, uint256 preJoinExitSupply)
         internal
-        returns (uint256)
+        view
+        returns (uint256, uint256)
     {
-        if (_exemptFromYieldFees) return 0;
+        if (_exemptFromYieldFees) return (0, 0);
 
         uint256 athRateProduct = _athRateProduct;
         uint256 rateProduct = _getRateProduct(normalizedWeights);
 
-        // Initialise `_athRateProduct`. This will occur on the first join/exit after Pool initialisation.
-        // Not initialising this here properly will cause all joins/exits to revert.
-        if (athRateProduct == 0) {
-            _athRateProduct = rateProduct;
-            return 0;
-        }
-
         // Only charge yield fees if we've exceeded the all time high of Pool value generated through yield.
```

Comment:
Should this big comment block be down in `_getRateProduct`?

Reply:
I don't think so. It's more of an explanation of how the yield fees mechanism works and so *why* we use those values. I think putting it in `_getRateProduct` would be hiding motivation inside implementation.

Reply:
Makes sense - maybe add a note that the logic described is implemented in `_getRateProduct`?

Reply:
That's good; for a second it looked wrong because the comment says invariant but it's about rate products. This clarifies that they're equivalent.